### PR TITLE
Wire dependency graph into JSON emission and --graph text rendering

### PR DESCRIPTION
**Pull request by JMBattista**
## Source

- Spec: [`smithy-status-skill.spec.md`](specs/2026-04-12-004-smithy-status-skill/smithy-status-skill.spec.md) (US10)
- Tasks: [`10-visualize-dependency-graph-for-parallel-work.tasks.md`](specs/2026-04-12-004-smithy-status-skill/10-visualize-dependency-graph-for-parallel-work.tasks.md)

## Slice

**Slice 3 of US10** — *Wire the Graph into JSON Emission and `--graph` Text Rendering*

`smithy status --format json` now emits a fully-populated `graph` object unconditionally, and `smithy status --graph` renders topological layers as text with done-layer collapsing (respecting `--all`) and a cycle-warning fallback. A new pure `src/status/renderGraph.ts` module owns the text layout; `statusAction` orchestrates the calls.

## Addresses

- **FR-026**, **FR-027**
- **AS 10.1** (layered output with US1/US4 in Layer 0)
- **AS 10.3** (cycle fallback with warning, no crash)
- **AS 10.4** (done-layer collapsing in default mode; `--all` expands)
- **AS 10.5** (JSON `graph` populated unconditionally with `node_ids`, `cycles`, `dangling_refs`)
- **AS 10.6** (dangling-ref warnings)

## Tasks completed

- [x] **Implement `renderGraph` in a new `src/status/renderGraph.ts` module** — pure function returning a layered text view; default-mode done-layer collapsing; cycle-warning fallback with flat listing of non-cyclic nodes; trailing `Dangling refs:` block; UTF-8 / ASCII glyph support; re-exported from `src/status/index.ts`. 16 unit tests covering AS 10.1, 10.3, 10.4, 10.6, purity, empty-graph, ASCII fallback, default theme.
- [x] **Wire `buildDependencyGraph` and `renderGraph` into `statusAction`** — graph built once per invocation against the **pre-filter** record set per SD-010 (so `--status` / `--type` never narrow the JSON `graph` shape); JSON payload's `graph` populated unconditionally, replacing the zero-value stub; `--graph` text mode prints the summary header and routes through `renderGraph(graph, { theme, all })`; default text path is untouched. Updated `StatusOptions.graph` JSDoc, `StatusJsonPayload` JSDoc, the module-level comment block, and the `--graph` Commander help text. 9 new integration tests covering AS 10.1, 10.3, 10.4, 10.5, 10.6 and the SD-010 pre-filter behavior.
- [x] **Audit `node_ids` vs `ids` across code and flag the spec-text drift** — confirmed `node_ids` is canonical end-to-end (`DependencyGraph` type, `buildDependencyGraph` output, `renderGraph` consumption, JSON payload). The only `.ids` reference under `src/status/` or `src/commands/status.ts` is in the SD-012 JSDoc commentary itself. SD-012 spec-prose drift surfaced for follow-up — see "PR notes" below.

## Review

`smithy-implementation-review` returned 5 findings.

**Auto-fix applied (1 Important x High)**:
- Missing `--graph` integration tests for the empty-repo and no-match-filter friendly-hint paths. Added two regression tests inside `statusAction --graph integration (US10 Slice 3)`. Commit: `a0b46b1 review: cover --graph empty-repo and no-match-filter friendly hints`.

**Notes for reviewer (4 Minor)**:
- `Layer N: DONE (M items)` separator inconsistency — the collapsed-layer heading uses a colon, while the Layer 0 "ready to work" heading and the contracts section 1 example (line 71) use an em-dash. Implementation matches the task wording (line 106 of the tasks file specifies the colon form). Worth reconciling either the contracts example to the colon form or the renderer to the em-dash form. Tracked under SD-006.
- Per-node lines drop the `(blocked by <ids>)` suffix and bracketed status text — the contracts section 1 example shows the bracketed `[not started]` form with a `(blocked by US1)` suffix; the implementation emits the icon-glyph form without either. The slice goal called this OPTIONAL and asked it be flagged if missing. Decision needed: extend the renderer or align the contracts example to the icon-glyph rendering. Tracked under SD-006.
- Cycle warning + dangling refs use `theme.paint.error` (red) — both are non-fatal warnings (scanner exits 0); painting them red reads as a hard error. Could be demoted to a `theme.paint.warning` (yellow) helper or `theme.paint.dim`. Polish-only.
- Missing dangling-refs + cycle combination test — the existing renderGraph unit tests cover each branch in isolation. A combined cycle + dangling-refs case and an all-dangling-no-layers case would tighten coverage.

## Documentation

- **Auto-fix applied** — `smithy-maid` flagged a stale future-tense JSDoc claim in `src/status/types.ts` (the module-level comment said the `--graph` text renderer would "arrive in Slice 3"). Replaced with past-tense language naming the new `renderGraph.ts` module. Commit: `ccf59f0 maid: refresh types.ts JSDoc — Slice 3 work is shipped`.

**Documentation Notes**:
- **SD-012 spec-prose drift remains open** — the spec document `smithy-status-skill.spec.md` AS 10.5 (line 206) still reads `layers: Array<{ layer: number, ids: string[] }>`. Code, data-model, and contracts JSON example all use `node_ids`. Per the slice instruction to NOT silently edit the spec, this PR leaves the spec prose untouched and surfaces the discrepancy here. Recommended follow-up: edit `specs/2026-04-12-004-smithy-status-skill/smithy-status-skill.spec.md:206` to change `ids: string[]` to `node_ids: string[]` and mark SD-012 resolved.

## Validation

| Command | Result |
|---|---|
| `npm run build` | clean — `dist/cli.js` 112.19 KB ESM |
| `npm run typecheck` | clean (no errors) |
| `npm test` | **664 / 664 passing** across 23 test files (up from 652 pre-slice) |

The newly-added tests cover AS 10.1, AS 10.3, AS 10.4, AS 10.5, AS 10.6, the SD-010 pre-filter behavior for the JSON `graph`, the SD-012 `node_ids` field-name enforcement, the empty-repo and no-match-filter friendly-hint paths under `--graph`, and a regression test that the default text path (no `--graph`) still routes through `renderTree`.

### CODE REVIEW

File: src/status/renderGraph.ts [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+  // `graph.layers`. Cyclic nodes are excluded from every layer entry by
+  // the builder, so iterating layers gives us exactly the set of
+  // non-cyclic nodes in topological order.
+  const flatIds: string[] = [];
+  for (const layer of graph.layers) {
+    for (const id of layer.node_ids) {
+      flatIds.push(id);
+    }
+  }
+  if (flatIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/status/renderGraph.ts]: Fixed in 43b7a6c: confirmed and fixed. `buildDependencyGraph` produces three node classes — in `graph.layers` (acyclic), in `graph.cycles` (cyclic), or in `graph.nodes` only (non-cyclic but downstream of a cycle, so Kahn's never reached zero in-degree). The flat fallback was iterating `graph.layers` only, silently dropping the third class. `formatCycleFallback` now sweeps `graph.layers` first (preserving topological order) then `graph.nodes`, deduplicating against an explicit `cyclicIds` set so cyclic nodes still surface only on the `Cycle:` line. Added a regression test with a US1 ↔ US2 cycle plus a US3 → US1 downstream dependent that explicitly asserts US3 was outside both `graph.layers` and `graph.cycles` in the builder output, then surfaces under `Nodes (flat fallback):`.

File: src/commands/status.ts [hunk 2]
```diff
@@ -210,10 +235,13 @@ export function statusAction(opts: StatusOptions = {}): void {
...
-  // for signature symmetry. `summarize(records)` stays above the
-  // filter call so `ScanSummary` / the header remain aggregate over
-  // the full scan per SD-010.
+  // for signature symmetry. `summarize(records)` and
+  // `buildDependencyGraph(records)` both stay above the filter call so
+  // the `ScanSummary` / header AND the `DependencyGraph` remain
+  // aggregate over the full scan per SD-010 — `--status` / `--type`
+  // narrow the rendered tree view but never the graph shape.
   const summary = summarize(records);
+  const graph = buildDependencyGraph(records);
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/commands/status.ts]: Not addressed: this contradicts the slice's explicit acceptance criterion. The Slice 3 task spec mandates the pre-filter record set: *"`buildDependencyGraph` is called exactly once per invocation (not once per consumer surface); the **pre-filter record set is used** so graph shape does not shift when `--status` / `--type` narrow the tree view"* (see [`10-visualize-dependency-graph-for-parallel-work.tasks.md` line 120](https://github.com/Balexda/SmithyCLI/blob/feature/status-forge-10-03/specs/2026-04-12-004-smithy-status-skill/10-visualize-dependency-graph-for-parallel-work.tasks.md#L120)).

This was tracked as **SD-010** (open spec debt) precisely because it conflicts with **SD-008** in the spec — which was resolved 2026-04-19 as "filters apply BEFORE layering". SD-010 explicitly notes the conflict: *"leaves open whether a future `--graph --status in-progress` should project or prune the graph. Closely related to inherited SD-008."* The slice deliberately punted that decision and shipped pre-filter behavior, with two integration tests locking it down (SD-010 pre-filter test for the JSON `graph` payload, plus the no-match-filter friendly-hint path under `--graph`).

This is a real spec design discussion, not a Slice 3 bug. The right next step is a follow-up that picks one rule for both surfaces — either re-resolve SD-008 to match SD-010 (keep current behavior) or re-resolve SD-010 to match SD-008 (re-build the graph from the filtered record set in a future slice). Changing the implementation here would violate the Slice 3 spec and break those two regression tests.

Flagging this thread for the spec author to drive.

File: src/status/renderGraph.ts [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+/**
+ * Format one layer block: heading line plus, when expanded, one tree-
+ * connector line per actionable member.
+ *
+ * Default mode (`all === false`) hides members with `status === 'done'`
+ * from the visible listing and tacks a `, N done hidden` suffix on the
+ * heading so the dropped work is still accounted for. When every
+ * member of the layer is `done` (zero actionable remainder), the whole
+ * layer collapses to the uniform `Layer N: DONE (M items)` line with
+ * no member listing.
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/status/renderGraph.ts]: Fixed in 075d0f4: refreshed the `formatLayerBlock` JSDoc to describe the actual current behavior (returns the empty string for fully-done layers, caller omits the layer entirely). Adopted the suggested wording verbatim.

File: src/commands/status.ts [hunk 4]
```diff
@@ -279,6 +304,41 @@ export function statusAction(opts: StatusOptions = {}): void {
     all: opts.all === true,
   });
   const topNextAction = pickTopNextAction(tree);
+
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/commands/status.ts:293]: Fixed in 9bdc76c: reordered `statusAction` so the `--graph` branch runs BEFORE the default-path tree pipeline. `buildTree` / `collapseTree` / `pickTopNextAction` now only execute when the default text view is actually about to render — the `--graph` branch replaces them outright, so they no longer pay for unused work.

File: src/commands/status.ts [hunk 5]
```diff
@@ -210,10 +235,13 @@ export function statusAction(opts: StatusOptions = {}): void {
...
-  // filter call so `ScanSummary` / the header remain aggregate over
-  // the full scan per SD-010.
+  // for signature symmetry. `summarize(records)` and
+  // `buildDependencyGraph(records)` both stay above the filter call so
+  // the `ScanSummary` / header AND the `DependencyGraph` remain
+  // aggregate over the full scan per SD-010 — `--status` / `--type`
+  // narrow the rendered tree view but never the graph shape.
   const summary = summarize(records);
+  const graph = buildDependencyGraph(records);
   // Build a sparse options object — `exactOptionalPropertyTypes` in
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/commands/status.ts:256]: Fixed in 9bdc76c: `buildDependencyGraph(records)` now lives behind a `getGraph()` closure that caches on first read. The JSON branch and the `--graph` text branch each call it lazily; the default text path never does. The two consumers that DO need the graph still share a single build (no double work). Pre-filter record set per SD-010 is preserved.

File: src/commands/status.test.ts [hunk 6]
```diff
@@ -453,3 +470,439 @@ describe('StatusJsonPayload.graph type wiring (US10 Slice 1)', () => {
...
+    );
+  }
+
+  /**
+   * Fixture where every spec row + tasks-file row in the graph rolls
+   * up to `done`. Both spec rows have all-checked tasks files, so the
+   * spec itself rolls up to `done`, and every node in the graph
+   * carries `status: 'done'`. In default mode every layer therefore
+   * collapses to `Layer N: DONE (M items)`; under `--all`, every
+   * layer expands and every node id surfaces.
```
[the remark this answers is not in the document]
Comment (JMBattista) [reply src/commands/status.test.ts]: Fixed in 075d0f4: refreshed the `writeAllDoneFixture` comment to match the assertions (default mode omits fully-done layers entirely; `--all` shows everything). Adopted the suggested wording verbatim.
